build_tools/python/cmake_builder/rules.py

```diff
@@ -0,0 +1,149 @@
+## Copyright 2022 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+"""Helpers that build CMake rules."""
+
+from typing import List, Optional, Sequence
+
+INDENT_SPACES = " " * 2
+
+
+def _get_string_list(values: List[str], quote: bool = True) -> List[str]:
+  if quote:
+    values = [f"\"{value}\"" for value in values]
+  return values
+
+
+def _get_block_body(body: List[str]) -> List[str]:
+  return [INDENT_SPACES + line for line in body]
+
+
+def _get_string_arg_block(keyword: str,
+                          value: Optional[str],
+                          quote: bool = True) -> List[str]:
+  if value is None:
+    return []
+  if quote:
+    value = f"\"{value}\""
+  return [keyword] + _get_block_body([value])
+
+
+def _get_string_list_arg_block(keyword: str,
+                               values: List[str],
+                               quote: bool = True) -> List[str]:
+  body = _get_string_list(values, quote)
+  if len(body) == 0:
+    return []
+  return [keyword] + _get_block_body(body)
+
+
+def _get_option_arg_block(keyword: str, value: Optional[bool]) -> List[str]:
+  if value is True:
+    return [keyword]
+  return []
+
+
+def _build_call_rule(rule_name: str,
+                     parameter_blocks: Sequence[List[str]]) -> List[str]:
+  output = [f"{rule_name}("]
+  for block in parameter_blocks:
+    if len(block) == 0:
+      continue
+    output.extend(_get_block_body(block))
+  output.append(")")
+  return output
+
+
+def _convert_block_to_string(block: List[str]) -> str:
+  return "\n".join(block) + "\n"
```

Comment:
Man it really *seems* like there should be a method for this using a terminator rather than a separator. I think this is going to re-allocate your whole string, which kind of sucks. If performance becomes a problem, you could add a newline to the last line before doing the join

```suggestion
  block[-1] += "\n"
  return "\n".join(block)
```

I guess rules aren't going to be that big though

Reply:
I decided to use `"\n".join(block + [""])` which only needs to copy the list. It just personally looks better to me.